File: backend/src/models/Comment.js
const mongoose = require('mongoose');

const commentSchema = new mongoose.Schema({
  // Media Info
  mediaId: {
    type: Number,
    required: true,
    index: true
  },
  mediaType: {
    type: String,
    required: true,
    enum: ['movie', 'tv', 'trending', 'upcoming']
  },

  // User Info
  userId: {
    type: String,
    required: true,
    index: true
  },
  userName: {
    type: String,
    required: true
  },
  userPhotoUrl: {
    type: String,
    default: 'https://www.pngitem.com/pimgs/m/146-1468479_my-profile-icon-blank-profile-picture-circle-hd.png'
  },

  // Comment Content
  text: {
    type: String,
    required: true,
    maxlength: 2000
  },

  // Parent Comment (for replies/threading)
  parentCommentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Comment',
    default: null
  },

  // Engagement
  likes: [{
    userId: String,
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  likesCount: {
    type: Number,
    default: 0
  },

  // Reply Count (for top-level comments)
  replyCount: {
    type: Number,
    default: 0
  },

  // Timestamps
  createdAt: {
    type: Date,
    default: Date.now,
    index: true
  },
  updatedAt: {
    type: Date,
    default: Date.now
  },

  // Soft Delete
  isDeleted: {
    type: Boolean,
    default: false
  }
});

// Indexes for performance
commentSchema.index({ mediaId: 1, mediaType: 1, createdAt: -1 });
commentSchema.index({ parentCommentId: 1, createdAt: -1 });
commentSchema.index({ userId: 1, createdAt: -1 });

// Virtual for replies (if needed)
commentSchema.virtual('replies', {
  ref: 'Comment',
  localField: '_id',
  foreignField: 'parentCommentId'
});

// Increment parent's reply count
commentSchema.pre('save', async function(next) {
  if (this.isNew && this.parentCommentId) {
    await mongoose.model('Comment').findByIdAndUpdate(
      this.parentCommentId,
      { $inc: { replyCount: 1 } }
    );
  }
  next();
});

// Update likesCount when likes array changes
commentSchema.pre('save', function(next) {
  this.likesCount = this.likes.length;
  next();
});

module.exports = mongoose.model('Comment', commentSchema);

File: lib/models/comment.dart
class Comment {
  final String id;
  final int mediaId;
  final String mediaType; // 'movie' or 'tv'
  final String userId;
  final String userName;
  final String? userPhotoUrl;
  final String text;
  final String? parentCommentId; // For replies
  final List<String> likes;
  final int likesCount;
  final int replyCount;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isDeleted;

  Comment({
    required this.id,
    required this.mediaId,
    required this.mediaType,
    required this.userId,
    required this.userName,
    this.userPhotoUrl,
    required this.text,
    this.parentCommentId,
    required this.likes,
    required this.likesCount,
    required this.replyCount,
    required this.createdAt,
    required this.updatedAt,
    this.isDeleted = false,
  });

  // Factory constructor from JSON
  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['_id'] ?? '',
      mediaId: json['mediaId'] ?? 0,
      mediaType: json['mediaType'] ?? 'movie',
      userId: json['userId'] ?? '',
      userName: json['userName'] ?? 'Anonymous',
      userPhotoUrl: json['userPhotoUrl'],
      text: json['text'] ?? '',
      parentCommentId: json['parentCommentId'],
      likes: List<String>.from(json['likes']?.map((l) => l['userId'].toString()) ?? []),
      likesCount: json['likesCount'] ?? 0,
      replyCount: json['replyCount'] ?? 0,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
      isDeleted: json['isDeleted'] ?? false,
    );
  }

  // Convert to JSON
  Map<String, dynamic> toJson() {
    return {
      'mediaId': mediaId,
      'mediaType': mediaType,
      'userId': userId,
      'userName': userName,
      'userPhotoUrl': userPhotoUrl,
      'text': text,
      'parentCommentId': parentCommentId,
    };
  }

  // Check if current user liked this comment
  bool isLikedByUser(String userId) {
    return likes.contains(userId);
  }

  // Copy with method for immutability
  Comment copyWith({
    String? id,
    int? mediaId,
    String? mediaType,
    String? userId,
    String? userName,
    String? userPhotoUrl,
    String? text,
    String? parentCommentId,
    List<String>? likes,
    int? likesCount,
    int? replyCount,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isDeleted,
  }) {
    return Comment(
      id: id ?? this.id,
      mediaId: mediaId ?? this.mediaId,
      mediaType: mediaType ?? this.mediaType,
      userId: userId ?? this.userId,
      userName: userName ?? this.userName,
      userPhotoUrl: userPhotoUrl ?? this.userPhotoUrl,
      text: text ?? this.text,
      parentCommentId: parentCommentId ?? this.parentCommentId,
      likes: likes ?? this.likes,
      likesCount: likesCount ?? this.likesCount,
      replyCount: replyCount ?? this.replyCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isDeleted: isDeleted ?? this.isDeleted,
    );
  }
}

class CommentStats {
  final int totalComments;
  final int topLevelComments;
  final int replies;

  CommentStats({
    required this.totalComments,
    required this.topLevelComments,
    required this.replies,
  });

  factory CommentStats.fromJson(Map<String, dynamic> json) {
    return CommentStats(
      totalComments: json['totalComments'] ?? 0,
      topLevelComments: json['topLevelComments'] ?? 0,
      replies: json['replies'] ?? 0,
    );
  }
}
